### DISCUSSION TITLE
Unable to Load Abstract Single Table Inheritance Entities After Upgrade to Play-Ebean 8.3.0

### DISCUSSION THREAD

**Discussion by robby-macphail** [playframework/playframework#12663] [General]
I am upgrading a Java Play application to Play-Ebean 8.30 from 7.30, both of which, I believe, use the same Ebean version (15.1.0). The application has many abstract entity classes using single table inheritance strategy along with a discriminator. This pattern has worked well previously, but since the upgrade, Ebean is unable to load any of the entities inheriting the abstract class.

For example:

```java
@Entity
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
@DiscriminatorColumn(name = "my_type_field", discriminatorType = DiscriminatorType.STRING)
public abstract class MyAbstractClass extends AnotherAbstractClass { ... }
````

```java
@Entity
@DiscriminatorValue("FOO")
public class MyConcreteClass extends MyAbstractClass { ... }
```

Attempting to load the entity results in:
```
Caused by: jakarta.persistence.PersistenceException: java.lang.UnsupportedOperationException: cannot create entity bean for abstract entity MyAbstractClass
at io.ebeaninternal.server.deploy.BeanDescriptor.createReference(BeanDescriptor.java:1521)
at io.ebeaninternal.server.deploy.BeanDescriptor.contextRef(BeanDescriptor.java:1730)
at io.ebeaninternal.server.deploy.AssocOneHelp.read(AssocOneHelp.java:70)
at io.ebeaninternal.server.deploy.AssocOneHelp.readSet(AssocOneHelp.java:81)
at io.ebeaninternal.server.deploy.BeanPropertyAssocOne.readSet(BeanPropertyAssocOne.java:624)
at io.ebeaninternal.server.query.SqlBeanLoad.load(SqlBeanLoad.java:63)
```
Has anything changed in recent Ebean versions to disallow using abstract classes with the single table inheritance?

**Comment by mkurz** [comment]
> to Play-Ebean 8.30 from 7.30

I guess you mean 8.3.0 and 7.3.0. So you are upgrading from Play 2.9 to 3.0?

There is absolutely no difference except the namespace switch and depending on Play 3.0 libraries instead of 2.9 ones. You can see that here:
https://github.com/playframework/play-ebean/compare/7.3.0..8.3.0

**Reply by robby-macphail** [reply]
Sorry, yes I meant 8.30 and 7.30. Sorry for the typo.

I see what's going on now. We are upgrading to Play 3.0.3 from 2.9.1 and from play-ebean 7.0.0. Play-ebean 7.0.0 still used Ebean 13. Now that we are moving to play-ebean 8.3.0, that bumps us up to Ebean 15.

I learned from the Ebean forum that [Ebean 15 will only support Jakarta and not javax.persistence](https://github.com/ebean-orm/ebean/releases/tag/15.0.0-RC) functionality, meaning it no longer supports single table inheritance. This is a significant breaking change for us.

Are there any plans for Play 3.0 to support Ebean 14? My understanding is that 14 and 15 are forks that will both be maintained for a period of time.

**Reply by mkurz** [reply]
> My understanding is that 14 and 15 are forks that will both be maintained for a period of time.

https://github.com/ebean-orm/ebean/releases/tag/15.0.0-RC states 

> _The expectation is that we will release 14.x and 15.x releases for some time in parallel (at least 1 year). Although we discussed the removal of these features some folks might not have picked that up until they see 15.x releases and maybe read these release notes so let's see how that goes._

Inheritance support was removed in
- https://github.com/ebean-orm/ebean/pull/3211

> Are there any plans for Play 3.0 to support Ebean 14? 

Hmm... We now ship with latest ebean 15.x versions. However you can downgrade to 14.x easily yourself. This will work because we do not rely on `javax.persistence`/`jakarta.persistence` annotations. Just add following dependency overrides to your project:

In `build.sbt` (based on [this code](https://github.com/playframework/play-ebean/blob/8.3.0/project/Dependencies.scala#L24-L26)):
```sbt
val ebean          = "14.3.0"
dependencyOverrides ++= Seq(
  ("io.ebean" % "ebean"               % s"$ebean-javax").excludeAll(ExclusionRule("com.fasterxml.jackson.core")),
  "io.ebean"  % "ebean-ddl-generator" % s"$ebean-javax",
  "io.ebean"  % "ebean-agent"         % ebean,
)
```

In your `project/plugins.sbt` where you already should have `addSbtPlugin("org.playframework" % "sbt-play-ebean" % "8.3.0")` (based on [this code](https://github.com/playframework/play-ebean/blob/8.3.0/project/Dependencies.scala#L38-L39)):
```sbt
val ebean = "14.3.0"
dependencyOverrides ++= Seq(
  "io.ebean" % "ebean"       % s"$ebean-javax",
  "io.ebean" % "ebean-agent" % ebean,
)
```

You should now have only ebean 14.3.0 dependency on your classpath, you can check that by running:
```sbt
show runtime:fullClasspath

# and the same for the plugins
reload plugins
show runtime:fullClasspath
```

Let me know if this works for you.

I recommend to either move away from the ebean inheritance feature or discuss with the maintainer here: 
- https://github.com/ebean-orm/ebean/discussions/3112
